### PR TITLE
Introduce API 1.30 (CPLib-based CENC PAVP support)

### PR DESCRIPTION
**Pull request by dmitrybrazhkin**
Signed-off-by: Dmitry Brazhkin <dmitry.brazhkin@intel.com>

### CODE REVIEW

File: _studio/shared/umc/codec/h264_dec/src/umc_h264_mfx_supplier.cpp [hunk 1]
```diff
@@ -1471,11 +1489,21 @@ mfxStatus MFX_Utility::Query(VideoCORE *core, mfxVideoParam *in, mfxVideoParam *
...
         return false;
 
+#if defined (MFX_ENABLE_CPLIB)
+    if (in->Protected)
+    {
+        if (type == MFX_HW_UNKNOWN || !IS_PROTECTION_CENC(in->Protected))
+            return false;
+    }
+#else
+    (void)type;
```
Comment (dvrogozh) [inline _studio/shared/umc/codec/h264_dec/src/umc_h264_mfx_supplier.cpp]: no need in #else. use unconditionally in the beginning of the function.

File: _studio/shared/umc/codec/h265_dec/src/umc_h265_mfx_utils.cpp [hunk 2]
```diff
@@ -922,11 +936,21 @@ mfxStatus Query_H265(VideoCORE *core, mfxVideoParam *in, mfxVideoParam *out, eMF
...
         return false;
 
+#if defined (MFX_ENABLE_CPLIB)
+    if (in->Protected)
+    {
+        if (type == MFX_HW_UNKNOWN || !IS_PROTECTION_CENC(in->Protected))
+            return false;
+    }
+#else
+    (void)type;
```
Comment (dvrogozh) [inline _studio/shared/umc/codec/h265_dec/src/umc_h265_mfx_utils.cpp]: no need in #else. use unconditionally in the beginning of the function

File: _studio/shared/src/libmfx_core_vaapi.cpp [hunk 3]
```diff
@@ -424,6 +424,22 @@ VAAPIVideoCORE::GetHandle(
     MFX_CHECK_NULL_PTR1(handle);
     UMC::AutomaticUMCMutex guard(m_guard);
 
+#if defined (MFX_ENABLE_CPLIB)
+#if (MFX_VERSION >= 1030)
```
Comment (dvrogozh) [inline _studio/shared/src/libmfx_core_vaapi.cpp]: I suggest to drop check for MFX_VERSION here and below where it is paired with MFX_ENABLE_CPLIB. Just make sure that we define MFX_ENABLE_CPLIB only if MFX_VERSION>=1030.

File: _studio/shared/umc/codec/h264_dec/src/umc_h264_va_packer.cpp [hunk 4]
```diff
@@ -887,6 +897,156 @@ Status PackerVA::QueryStreamOut(H264DecoderFrame* pFrame)
     return UMC_OK;
 }
 
+#ifdef MFX_ENABLE_CPLIB
+
+PackerVA_CENC::PackerVA_CENC(VideoAccelerator * va, TaskSupplier * supplier)
```
Comment (dvrogozh) [inline _studio/shared/umc/codec/h264_dec/src/umc_h264_va_packer.cpp]: IMHO, having this in a separate file as we do for HEVC seems to be much better.

File: _studio/shared/include/mfx_cenc.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+#ifdef MFX_ENABLE_CPLIB
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define VACencStatusParameterBufferType    -3
+
+#define VA_ENCRYPTION_TYPE_CENC_CBC         0x00000002
+#define VA_ENCRYPTION_TYPE_CENC_CTR_LENGTH  0x00000004
```
Comment (dvrogozh) [inline _studio/shared/include/mfx_cenc.h]: hmmm. Why VAAPI does not hold these definitions? Is there any upstream plan or Android will have patched VAAPI?

File: _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp [hunk 6]
```diff
@@ -967,6 +970,12 @@ mfxStatus VideoDECODEH264::QueryIOSurfInternal(eMFXPlatform platform, eMFXHWType
         dpbSize = par->mfx.MaxDecFrameBuffering;
 
     mfxU32 numMin = dpbSize + 1 + asyncDepth;
+
+#ifdef MFX_ENABLE_CPLIB
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp]: Do we actually need to hide all that under MFX_ENABLE_CPLIB? I feel we can limit MFX_ENABLE_CPLIB usage to just few places where we really create some classes to handle protected content and maybe some checks in Init to return error if anyone tries to use protected if support is not available.

File: _studio/shared/include/mfx_utils.h [hunk 7]
```diff
@@ -121,6 +121,11 @@ bool LumaIsNull(const mfxFrameSurface1 * surf)
 #define SAFE_RELEASE(PTR)   { if (PTR) { PTR->Release(); PTR = NULL; } }
 #endif
 
+#ifdef MFX_ENABLE_CPLIB
+    #define IS_PROTECTION_CENC(val) (MFX_PROTECTION_CENC_WV_CLASSIC == (val) || MFX_PROTECTION_CENC_WV_GOOGLE_DASH == (val))
+    #define IS_PROTECTION_ANY(val) IS_PROTECTION_CENC(val)
```
Comment (dvrogozh) [inline _studio/shared/include/mfx_utils.h]: You might wish to use MFX_VERSION check here and have #else path to still define IS_* functions. In this way you might reduce number of MFX_ENABLE_CPLIB macro checks.

File: _studio/shared/umc/io/umc_va/src/umc_va_linux_protected.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+{
+    return m_protected;
+}
+
+void ProtectedVA::SetBitstream(mfxBitstream *bs)
+{
+    if (!bs)
+        return;
+
+#ifdef MFX_ENABLE_CPLIB
```
Comment (dvrogozh) [inline _studio/shared/umc/io/umc_va/src/umc_va_linux_protected.cpp]: so, ProtectedVA is not actually covered by MFX_ENABLE_CPLIB. You might be able to remove some of MFX_ENABLE_CPLIB in the code which operates over ProtectedVA like delete.

File: android/mfx_defs.mk [hunk 9]
```diff
@@ -39,6 +39,12 @@ endif
 # because there is used old version of LibVA 2.0
 ifneq ($(filter MFX_O MFX_O_MR1, $(MFX_ANDROID_VERSION)),)
   MFX_CFLAGS += -DMFX_VERSION=1026
+else
+  # CPLib PAVP implementation
+  # It requires minimum API version 1.30
+  MFX_CFLAGS += \
+    -DMFX_ENABLE_CPLIB \
+    -DMFX_VERSION=1030
```
Comment (dvrogozh) [inline android/mfx_defs.mk]: Add MFX_ENABLE_CPLIB into https://github.com/Intel-Media-SDK/MediaSDK/blob/master/mfxconfig.h.in and make sure we get `/* #undef MFX_ENABLE_CPLIB */` line in generated mfxconfig.h.

File: _studio/shared/umc/codec/h264_dec/src/umc_h264_mfx_supplier.cpp (hunk 1 above)
Comment (dmitrybrazhkin) [reply _studio/shared/umc/codec/h264_dec/src/umc_h264_mfx_supplier.cpp]: fixed

File: _studio/shared/umc/codec/h265_dec/src/umc_h265_mfx_utils.cpp (hunk 2 above)
Comment (dmitrybrazhkin) [reply _studio/shared/umc/codec/h265_dec/src/umc_h265_mfx_utils.cpp]: fixed

File: _studio/shared/src/libmfx_core_vaapi.cpp (hunk 3 above)
Comment (dmitrybrazhkin) [reply _studio/shared/src/libmfx_core_vaapi.cpp]: fixed

File: _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp (hunk 6 above)
Comment (dmitrybrazhkin) [reply _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp]: fixed

File: _studio/shared/include/mfx_utils.h (hunk 7 above)
Comment (dmitrybrazhkin) [reply _studio/shared/include/mfx_utils.h]: fixed

File: _studio/shared/umc/io/umc_va/src/umc_va_linux_protected.cpp (hunk 8 above)
Comment (dmitrybrazhkin) [reply _studio/shared/umc/io/umc_va/src/umc_va_linux_protected.cpp]: fixed

File: android/mfx_defs.mk (hunk 9 above)
Comment (dmitrybrazhkin) [reply android/mfx_defs.mk]: What is the syntax of mfxconfig.h.in? I see how to do "define" but it's not clear how to do "undef"

Comment (dvrogozh) [reply android/mfx_defs.mk]: You don't need to explicitly type anything for undef. cmake parses this .in file and it substitutes #cmakedefine YOUR_DEF with either  #define YOUR_DEF.... if YOUR_DEF was defined on cmake level and with /* #undef YOUR_DEF */ if it was not.

Comment (dmitrybrazhkin) [reply android/mfx_defs.mk]: I've added MFX_ENABLE_CPLIB into https://github.com/Intel-Media-SDK/MediaSDK/blob/master/mfxconfig.h.in and got /* #undef MFX_ENABLE_CPLIB */ line in generated mfxconfig.h. Do you want me to upload this change, or it was just a check?

File: _studio/shared/umc/core/umc/include/umc_va_base.h [hunk 10]
```diff
@@ -188,6 +188,7 @@ class VideoAccelerator
         m_Profile(UNKNOWN),
         m_Platform(VA_UNKNOWN_PLATFORM),
         m_HWPlatform(MFX_HW_UNKNOWN),
+        m_protectedVA(0),
```
Comment (dvrogozh) [inline _studio/shared/umc/core/umc/include/umc_va_base.h]: nullptr

File: _studio/shared/umc/io/umc_va/src/umc_va_linux.cpp [hunk 11]
```diff
@@ -560,6 +589,8 @@ Status LinuxVideoAccelerator::Close(void)
         m_dpy = NULL;
     }
 
+    delete m_protectedVA;
+    m_protectedVA = 0;
```
Comment (dvrogozh) [inline _studio/shared/umc/io/umc_va/src/umc_va_linux.cpp]: nullptr

File: android/mfx_defs.mk (hunk 9 above)
Comment (dvrogozh) [reply android/mfx_defs.mk]: Yes, upload please. Better to have this explicit.

File: _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp [hunk 12]
```diff
@@ -34,6 +34,9 @@
 #include "vm_sys_info.h"
 
 #include "umc_h264_va_supplier.h"
+#if defined(MFX_ENABLE_CPLIB)
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp]: no need in protection here, I think

File: _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp [hunk 13]
```diff
@@ -1197,6 +1204,15 @@ mfxStatus VideoDECODEH264::DecodeFrameCheck(mfxBitstream *bs, mfxFrameSurface1 *
     sts = MFX_ERR_UNDEFINED_BEHAVIOR;
 
 #if defined (MFX_VA_LINUX)
+#if defined(MFX_ENABLE_CPLIB)
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp]: no need in protection here, I think

File: _studio/mfx_lib/decode/h265/src/mfx_h265_dec_decode.cpp [hunk 14]
```diff
@@ -32,6 +32,9 @@
 #include "vm_sys_info.h"
 
 #include "umc_h265_va_supplier.h"
+#if defined(MFX_ENABLE_CPLIB)
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/h265/src/mfx_h265_dec_decode.cpp]: no need in protection here, I think

File: _studio/mfx_lib/decode/h265/src/mfx_h265_dec_decode.cpp [hunk 15]
```diff
@@ -1001,6 +1004,15 @@ mfxStatus VideoDECODEH265::DecodeFrameCheck(mfxBitstream *bs, mfxFrameSurface1 *
 
     sts = MFX_ERR_UNDEFINED_BEHAVIOR;
 
+#if defined(MFX_ENABLE_CPLIB)
```
Comment (dvrogozh) [inline _studio/mfx_lib/decode/h265/src/mfx_h265_dec_decode.cpp]: no need in protection here, I think

File: _studio/shared/umc/core/umc/include/umc_va_base.h (hunk 10 above)
Comment (dmitrybrazhkin) [reply _studio/shared/umc/core/umc/include/umc_va_base.h]: done

File: _studio/shared/umc/io/umc_va/src/umc_va_linux.cpp (hunk 11 above)
Comment (dmitrybrazhkin) [reply _studio/shared/umc/io/umc_va/src/umc_va_linux.cpp]: done

File: _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp (hunk 12 above)
Comment (dmitrybrazhkin) [reply _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp]: done

File: _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp (hunk 13 above)
Comment (dmitrybrazhkin) [reply _studio/mfx_lib/decode/h264/src/mfx_h264_dec_decode.cpp]: done

File: _studio/mfx_lib/decode/h265/src/mfx_h265_dec_decode.cpp (hunk 14 above)
Comment (dmitrybrazhkin) [reply _studio/mfx_lib/decode/h265/src/mfx_h265_dec_decode.cpp]: done

File: _studio/mfx_lib/decode/h265/src/mfx_h265_dec_decode.cpp (hunk 15 above)
Comment (dmitrybrazhkin) [reply _studio/mfx_lib/decode/h265/src/mfx_h265_dec_decode.cpp]: done

File: android/mfx_defs.mk (hunk 9 above)
Comment (dmitrybrazhkin) [reply android/mfx_defs.mk]: done

**Comment by dmitrybrazhkin** [comment]
> Please update documentation along with API change.

done

**Comment by dvrogozh** [comment]
windows build fails for some reason...

**Comment by dmitrybrazhkin** [comment]
> windows build fails for some reason...

Looks like infrastructure issue. I've committed once again to restart buildbot

**Pull request merged**